controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -57,10 +58,88 @@ type EndpointsController struct {
 	queue      workqueue.RateLimitingInterface
 	reconciler *endpointsReconciler
 	stop       chan struct{}
-	isLeader   atomic.Bool
 
 	lec leaderelection.LeaderElectionConfig
-	sync.RWMutex
+	informerHandlers
+}
+
+// informerHandlers holds handles to callbacks that have been registered with
+// the API Server client's informers.
+//
+// These callbacks will be registered when a controller is elected as leader,
+// and de-registered when the lease is lost.
+type informerHandlers struct {
+	ewHandle  cache.ResourceEventHandlerRegistration
+	esHandle  cache.ResourceEventHandlerRegistration
+	svcHandle cache.ResourceEventHandlerRegistration
+
+	sync.Mutex
+}
+
+// addHandlers will register a set of callbacks with the different informers
+// needed to synchronise endpoint state.
+func (ec *EndpointsController) addHandlers() error {
```

Comment:
There is going to be a problem here. Imagine the following sequence of events: 

1. Controller acquire lease.
2. Creates endpoints for a service and external workload
3. Drops lease removes callback
4. Meanwhile endpoints membership is modified by another controller
5. Our controllers aquires the lease
6. We get an ADD for the service, external workload,etc
7. We try to do a reconciliation, but the state of the endpoints tracker is stale because we never saw the DELETE
8. We try to requeue but our state never recovers and we drop the update.
9. We end up in an inconsistent state.

The problem here is that you are keeping state in the endpointsslice tracker that might change under you feet. If you look at the upstream implementation, you will notice that we start with a new endpoitnsslices tracker each time we aquire the lease. 

We need to do that here as well. Keeping this state around will get us into trouble.